### PR TITLE
662 site apportionment unit tests

### PR DESCRIPTION
**Pull request by jwestw**
## Pull Request submission

This set of unit tests tests the `count_unique_postcodes_in_col` function. 

`TestCreateNotnullMask` was created by Anne. 
`test_count_unique_postcodes_in_col` tests that the postcode count works as expected
`test_count_unqiue_postcodes_in_col_repetitive` tests that the function can handle repeated postcodes for the same and different references. 
`test_count_unqiue_postcodes_in_col_blank_postcodes` tests whether the function can handle different types of blanks, "", "  " (whitespace) and `None`
`test_count_unqiue_postcodes_in_col_nan_postcodes` tests if the function can handle different types of NaN values, including `str(np.nan)`, which may occur as we are picking up the data from another system. 
`test_count_unqiue_postcodes_in_col_missing_col` tests what would happen if the data was missing a key column, making sure that it raises an error.
`test_count_unqiue_postcodes_drop_duplicates` tests that the function will handle these repeats (dropping them during the process, but putting them back on the original data), with correct counts.
`test_count_unqiue_postcodes_drop_duplicates_with_missing_values` tests how the function behaves when there is repititious data but it has values missing from the row. This means that the entire row is not a duplicate (which `.drop_duplicates()` relies on, unless a subset is specified. 

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: tests/test_site_apportionment/test_site_apportionment.py [hunk 1]
```diff
@@ -58,3 +67,325 @@ def test_create_notnull_mask(self):
...
+    Returns:
+        pd.DataFrame: The created DataFrame.
+    """
+    sample_cols = ["reference", "period", "601", "headcount_dummy", "211_dummy"]
+
+    sample_data = [
+    [4990000000, 202212, "B27 2CD", 5663, 8855030],
+    [4990000000, 202212, "B27 2CD", 5663, 8855030], #repeated postcode
+    [4990000084, 202212, "B27 2CD", 8020, 5085659], # repeated postcode but for different ref
+    [4990000252, 202212, "UV1 2WX", 8020, 5085659],
```
Comment (JenCheshire) [inline tests/test_site_apportionment/test_site_apportionment.py]: References in the result_df and extpecetd_df are different so test failing

File: tests/test_site_apportionment/test_site_apportionment.py [hunk 2]
```diff
@@ -58,3 +67,325 @@ def test_create_notnull_mask(self):
...
+        [4990000000, 202212, "B27 2CD", 5663, 8855030, 1],
+        [4990000000, 202212, "B27 2CD", 5663, 8855030, 1], # repeated postcode but for the same ref
+        [4990000002, 202212, "B27 2CD", 8020, 5085659, 2], # postcode again repeated but for different ref
+        [4990000002, 202212, 'UV1 2WX', 8020, 5085659, 2] # postcode_count should be 2
+    ]
+
+    expected_df = pd.DataFrame(data=expected_data, columns=expected_cols)
+
+    # Check if the output is not as expected
+    with pytest.raises(AssertionError):
```
Comment (JenCheshire) [inline tests/test_site_apportionment/test_site_apportionment.py]: assert_frame_equal not working here, but not sure why.  If I replace lines 168/169 with assert_frame_equal(result_df, expected_df) it works correctly

File: tests/test_site_apportionment/test_site_apportionment.py [hunk 3]
```diff
@@ -58,3 +67,325 @@ def test_create_notnull_mask(self):
...
+    [4990000252, 202212, 7084, 7058606],
+    [4990000294, 202212, 3188, 1828472],
+    [4990000336, 202212, 62, 3262548],
+    [4990000378, 202212, 1180, 9348647]
+    ]
+
+    return pd.DataFrame(data=sample_data, columns=sample_cols)
+
+
+def test_count_unique_postcodes_in_col_missigng_col():
```
Comment (JenCheshire) [inline tests/test_site_apportionment/test_site_apportionment.py]: typo in function name (...missigng...)

File: tests/test_site_apportionment/test_site_apportionment.py [hunk 4]
```diff
@@ -58,3 +67,325 @@ def test_create_notnull_mask(self):
...
+    """Checks that an error is raised when the postcode column is missing."""
+
+    df = create_sample_missing_columns_df()
+
+    # Apply the function
+    with pytest.raises(KeyError):
```
Comment (JenCheshire) [inline tests/test_site_apportionment/test_site_apportionment.py:297]: This is throwing up a similar issue as test_count_unique_postcodes_in_col_repetitive, where I've changed the name of the col in the test to 601 and it is still passing

File: tests/test_site_apportionment/test_site_apportionment.py [hunk 5]
```diff
@@ -58,3 +67,325 @@ def test_create_notnull_mask(self):
...
+
+def create_sample_blank_postcodes_df():
+
+    sample_cols = ["reference", "period", "601", "headcount_dummy", "211_dummy"]
+
+    sample_data =    [
+        [4990000126, 202212, "IJ5 6KL", 3877, 5808144],
+        [4990000126, 202212, "",        3877, 5808144], #blank string as postcode
+        [4990000126, 202212, "   ",     3877, 5808144], # whitespace as postcode
+        [4990000126, 202212, None,      3877, 5808144]
```
Comment (JenCheshire) [inline tests/test_site_apportionment/test_site_apportionment.py:179]: To discuss: empty string is being treated as a postcode

**Comment by AnneONS** [comment]
Tests completed elsewhere

**Pull request closed without merging**